### ISSUE TITLE
Implementing SSLSocket and SSLSocketFactory

### ISSUE THREAD

**Issue by FD-** [google/j2objc#998]
I'm trying to implement an SSLSocketFactory to use within a j2objc project. I decided to use the [Secure Transport API](https://developer.apple.com/documentation/security/secure_transport). In theory, it seems like the API provides exactly what I was looking for: It offers a TLS layer that, by means of callbacks, can be used with any lower layer channel. I don't need any of the customisability, I just want to be able to secure a Socket over TLS/SSL to connect to an HTTPS server for TCP-level communication.

I created a WrappedSSLSocket class in Java, that just extends from the SSLSocket class and forwards all method calls to an underlying Socket.

<details><summary>WrappedSSLSocket.java</summary>
<p>

```
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetAddress;
import java.net.Socket;
import java.net.SocketAddress;
import java.net.SocketException;
import java.nio.channels.SocketChannel;

import javax.net.ssl.HandshakeCompletedListener;
import javax.net.ssl.SSLSession;
import javax.net.ssl.SSLSocket;

public class WrappedSSLSocket extends SSLSocket {
    protected Socket underlyingSocket;

    public WrappedSSLSocket(Socket underlying){
        this.underlyingSocket = underlying;
    }

    public Socket getUnderlyingSocket(){
        return underlyingSocket;
    }

    public void connect(SocketAddress endpoint) throws IOException {
        throw new RuntimeException("Stub!");
    }

    public void connect(SocketAddress endpoint, int timeout) throws IOException {
        underlyingSocket.connect(endpoint, timeout);
    }

    public void bind(SocketAddress bindpoint) throws IOException {
        underlyingSocket.bind(bindpoint);
    }

    public InetAddress getInetAddress() {
        return underlyingSocket.getInetAddress();
    }

    public InetAddress getLocalAddress() {
        return underlyingSocket.getLocalAddress();
    }

    public int getPort() {
        return underlyingSocket.getPort();
    }

    public int getLocalPort() {
        return underlyingSocket.getLocalPort();
    }

    public SocketAddress getRemoteSocketAddress() {
        return underlyingSocket.getRemoteSocketAddress();
    }

    public SocketAddress getLocalSocketAddress() {
        return underlyingSocket.getLocalSocketAddress();
    }

    public SocketChannel getChannel() {
        return underlyingSocket.getChannel();
    }

    public InputStream getInputStream() throws IOException {
        return underlyingSocket.getInputStream();
    }

    public OutputStream getOutputStream() throws IOException {
        return underlyingSocket.getOutputStream();
    }

    public void setTcpNoDelay(boolean on) throws SocketException {
        underlyingSocket.setTcpNoDelay(on);
    }

    public boolean getTcpNoDelay() throws SocketException {
        return underlyingSocket.getTcpNoDelay();
    }

    public void setSoLinger(boolean on, int linger) throws SocketException {
        underlyingSocket.setSoLinger(on, linger);
    }

    public int getSoLinger() throws SocketException {
        return underlyingSocket.getSoLinger();
    }

    public void sendUrgentData(int data) throws IOException {
        underlyingSocket.sendUrgentData(data);
    }

    public void setOOBInline(boolean on) throws SocketException {
        underlyingSocket.setOOBInline(on);
    }

    public boolean getOOBInline() throws SocketException {
        return underlyingSocket.getOOBInline();
    }

    public synchronized void setSoTimeout(int timeout) throws SocketException {
        underlyingSocket.setSoTimeout(timeout);
    }

    public synchronized int getSoTimeout() throws SocketException {
        return underlyingSocket.getSoTimeout();
    }

    public synchronized void setSendBufferSize(int size) throws SocketException {
        underlyingSocket.setSendBufferSize(size);
    }

    public synchronized int getSendBufferSize() throws SocketException {
        return underlyingSocket.getSendBufferSize();
    }

    public synchronized void setReceiveBufferSize(int size) throws SocketException {
        underlyingSocket.setSendBufferSize(size);
    }

    public synchronized int getReceiveBufferSize() throws SocketException {
        return underlyingSocket.getReceiveBufferSize();
    }

    public void setKeepAlive(boolean on) throws SocketException {
        underlyingSocket.setKeepAlive(on);
    }

    public boolean getKeepAlive() throws SocketException {
        return underlyingSocket.getKeepAlive();
    }

    public void setTrafficClass(int tc) throws SocketException {
        underlyingSocket.setTrafficClass(tc);
    }

    public int getTrafficClass() throws SocketException {
        return underlyingSocket.getTrafficClass();
    }

    public void setReuseAddress(boolean on) throws SocketException {
        underlyingSocket.setReuseAddress(on);
    }

    public boolean getReuseAddress() throws SocketException {
        return underlyingSocket.getReuseAddress();
    }

    public synchronized void close() throws IOException {
        underlyingSocket.close();
    }

    public void shutdownInput() throws IOException {
        underlyingSocket.shutdownInput();
    }

    public void shutdownOutput() throws IOException {
        underlyingSocket.shutdownOutput();
    }

    public String toString() {
        return "WrappedSocket (" + underlyingSocket.toString() + ")";
    }

    public boolean isConnected() {
        return underlyingSocket.isConnected();
    }

    public boolean isBound() {
        return underlyingSocket.isBound();
    }

    public boolean isClosed() {
        return underlyingSocket.isClosed();
    }

    public boolean isInputShutdown() {
        return underlyingSocket.isInputShutdown();
    }

    public boolean isOutputShutdown() {
        return underlyingSocket.isOutputShutdown();
    }

    public void setPerformancePreferences(int connectionTime, int latency, int bandwidth) {
        underlyingSocket.setPerformancePreferences(connectionTime, latency, bandwidth);
    }

    /* SSLSocket implementation */

    @Override
    public String[] getSupportedCipherSuites() {
        // Not implemented
        return new String[0];
    }

    @Override
    public String[] getEnabledCipherSuites() {
        // Not implemented
        return new String[0];
    }

    @Override
    public void setEnabledCipherSuites(String[] strings) {
        // Not implemented
    }

    @Override
    public String[] getSupportedProtocols() {
        // Not implemented
        return new String[0];
    }

    @Override
    public String[] getEnabledProtocols() {
        // Not implemented
        return new String[0];
    }

    @Override
    public void setEnabledProtocols(String[] strings) {
        // Not implemented
    }

    @Override
    public SSLSession getSession() {
        // Not implemented
        return null;
    }

    @Override
    public void addHandshakeCompletedListener(HandshakeCompletedListener handshakeCompletedListener) {
        // Not implemented
    }

    @Override
    public void removeHandshakeCompletedListener(HandshakeCompletedListener handshakeCompletedListener) {
        // Not implemented
    }

    @Override
    public void startHandshake() throws IOException {

    }

    @Override
    public void setUseClientMode(boolean b) {
        // Not implemented
    }

    @Override
    public boolean getUseClientMode() {
        return true;
    }

    @Override
    public void setNeedClientAuth(boolean b) {
        // Not implemented
    }

    @Override
    public boolean getNeedClientAuth() {
        // Not implemented
        return false;
    }

    @Override
    public void setWantClientAuth(boolean b) {
        // Not implemented
    }

    @Override
    public boolean getWantClientAuth() {
        // Not implemented
        return false;
    }

    @Override
    public void setEnableSessionCreation(boolean b) {
        // Not implemented
    }

    @Override
    public boolean getEnableSessionCreation() {
        // Not implemented
        return false;
    }

```

</p>
</details>
<p></p>

I then created a Swift class that extends from the j2objc-translated WrappedSSLSocket class. There, I create an SSLContext and set up its IO callbacks to read and write via the underlying Socket. From the `doHandshake` method, I call `SSLHandshake`.

<details><summary>iOSSSLSocketFactory.swift</summary>
<p>

```
import Foundation
import Security

class iOSSSLInputStream : JavaIoInputStream {
  unowned let sslSocket : iOSSSLSocket
  
  init(sslSocket : iOSSSLSocket) {
    self.sslSocket = sslSocket
    super.init()
  }
  
  override func read(with b: IOSByteArray!, with off: jint, with len: jint) -> jint {
    if (sslSocket.isClosed() || (sslSocket.underlyingSocket?.isClosed())!){
      ObjC.throwException(JavaIoIOException(nsString: "Cannot read from closed socket"))
    }
    
    if (sslSocket.isInputShutdown() || (sslSocket.underlyingSocket?.isInputShutdown())!){
      ObjC.throwException(JavaIoIOException(nsString: "Cannot read from shutdown socket"))
    }
    
    let unsafePointer = UnsafeMutableRawPointer(b.byteRef(at: UInt(off)))
    var actuallyRead : Int = 0
    let status = SSLRead(sslSocket.getSSLContext()!, unsafePointer!, Int(len), &actuallyRead)
    if (status == errSecSuccess){
      return jint(actuallyRead)
    } else {
      ObjC.throwException(JavaIoIOException(nsString: "Error reading from SSL: " + String(status)))
    }
    return jint(actuallyRead)
  }
  
  override func read(with b: IOSByteArray!) -> jint {
    return read(with: b, with: 0, with: b.length())
  }
  
  override func read() -> jint {
    let buffer = IOSByteArray.newArray(withLength: 1)
    let actuallyRead = read(with: buffer)
    if (actuallyRead == 1) {
      let resultByte : jbyte = (buffer?.byte(at: 0))!
      return jint(resultByte)
    } else {
      return actuallyRead
    }
  }
  
  override func close() {
    sslSocket.close()
  }
  
  override func available() -> jint {
    // Not supported
    return 0;
  }
  
  override func mark(with readlimit: jint) {
    // Not supported
  }
  
  override func markSupported() -> jboolean {
    return false
  }
  
  override func reset() {
    // Not supported
  }
  
  override func skip(withLong n: jlong) -> jlong {
    // Not supported
    return 0
  }
}

class iOSSSLOutputStream : JavaIoOutputStream {
  unowned let sslSocket : iOSSSLSocket
  
  init(sslSocket : iOSSSLSocket) {
    self.sslSocket = sslSocket
    super.init()
  }
  
  override func write(with b: IOSByteArray!, with off: jint, with len: jint) {
    if (sslSocket.isClosed() || (sslSocket.underlyingSocket?.isClosed())!){
      ObjC.throwException(JavaIoIOException(nsString: "Cannot write to closed socket"))
    }
    
    if (sslSocket.isOutputShutdown() || (sslSocket.underlyingSocket?.isOutputShutdown())!){
      ObjC.throwException(JavaIoIOException(nsString: "Cannot write to shutdown socket"))
    }
    
    var unsafePointer = UnsafeMutableRawPointer(b.byteRef(at: UInt(off)))
    var actuallyWrote : Int = 0
    var remaining : Int = Int(len);
    
    while remaining > 0 {
      let status = SSLWrite(sslSocket.getSSLContext()!, unsafePointer, remaining, &actuallyWrote)
      if (status == noErr){
        unsafePointer = unsafePointer?.advanced(by: actuallyWrote)
        remaining -= actuallyWrote
      } else {
        ObjC.throwException(JavaIoIOException(nsString: "Error writing to SSL: " + String(status)))
      }
    }
  }
  
  override func write(with b: IOSByteArray!) {
    write(with: b, with: 0, with: b.length())
  }
  
  override func write(with b: jint) {
    let buffer = IOSByteArray.newArray(withLength: 1)
    buffer?.replaceByte(at: 0, withByte: jbyte(b))
    write(with: buffer)
  }
  
  override func close() {
    sslSocket.close()
  }
  
  override func flush() {
    sslSocket.getUnderlyingSocket().getOutputStream().flush()
  }
}

class iOSSSLSocket : WrappedSSLSocket{
  var sslContext : SSLContext?
  var inputStream : iOSSSLInputStream?
  var outputStream : iOSSSLOutputStream?
  var underlyingSocket : JavaNetSocket?
  
  init(underlyingSocket: JavaNetSocket, hostName: String) {
    self.underlyingSocket = underlyingSocket
    self.sslContext = SSLCreateContext(nil, SSLProtocolSide.clientSide, SSLConnectionType.streamType)
  
    super.init(javaNetSocket: underlyingSocket)
    
    self.inputStream = iOSSSLInputStream(sslSocket: self)
    self.outputStream = iOSSSLOutputStream(sslSocket: self)
  
    var status = noErr
    status = SSLSetIOFuncs(sslContext!, sslReadCallback, sslWriteCallback)
    if (status != noErr) {
      ObjC.throwException(JavaIoIOException(nsString: "Error setting IO functions: " + String(status)))
    }
    
    let ref : SSLConnectionRef = UnsafeRawPointer(Unmanaged.passUnretained(self).toOpaque())
    status = SSLSetConnection(sslContext!, ref)
    if (status != noErr) {
      ObjC.throwException(JavaIoIOException(nsString: "Error setting connection data: " + String(status)))
    }
    
    status = SSLSetPeerDomainName(sslContext!, hostName, hostName.lengthOfBytes(using: String.Encoding.utf8))
    if (status != noErr) {
      ObjC.throwException(JavaIoIOException(nsString: "Error setting domain name: " + String(status)))
    }
  }
  
  override func startHandshake() {
    let status = SSLHandshake(sslContext!)
    if (status != noErr) {
      let cfDescription : CFString = SecCopyErrorMessageString(status, nil)!
      let description = cfDescription as String
      ObjC.throwException(JavaIoIOException(nsString: "Handshake error: " + String(status) + " - " + description))
    }
  }
  
  func getSSLContext() -> SSLContext?{
    return sslContext
  }
  
  var sslReadCallback : @convention(c) (SSLConnectionRef, UnsafeMutableRawPointer, UnsafeMutablePointer<Int>) -> OSStatus =
  {(connection: SSLConnectionRef, data: UnsafeMutableRawPointer, size: UnsafeMutablePointer<Int>) -> OSStatus in
    
    let socket = Unmanaged<iOSSSLSocket>.fromOpaque(connection).takeUnretainedValue()
    
    let javaByteBuffer = IOSByteArray.newArray(withLength: UInt(size.pointee))
    var result = noErr
    
    do {
      try ObjC.catchException {
        let actuallyRead = socket.getUnderlyingSocket().getInputStream().read(with: javaByteBuffer)
        if (actuallyRead >= 0) {
          let jbytePointer = data.bindMemory(to: jbyte.self, capacity: size.pointee)
          javaByteBuffer?.getBytes(jbytePointer, length:UInt(actuallyRead))
          size.pointee = Int(actuallyRead)
        } else {
          result = errSSLBadCert
        }
      }
    } catch {
      result = errSSLClosedAbort
    }
    return result
  }
  
  var sslWriteCallback : @convention(c) (SSLConnectionRef, UnsafeRawPointer, UnsafeMutablePointer<Int>) -> OSStatus =
  {(connection: SSLConnectionRef, data: UnsafeRawPointer, size: UnsafeMutablePointer<Int>) -> OSStatus in

    let socket = Unmanaged<iOSSSLSocket>.fromOpaque(connection).takeUnretainedValue()

    let jbytePointer = data.bindMemory(to: jbyte.self, capacity: size.pointee)
    let javaByteBuffer = IOSByteArray.newArray(withBytes:jbytePointer, count: UInt(size.pointee))
    
    do {
      try ObjC.catchException {
        socket.getUnderlyingSocket().getOutputStream().write(with: javaByteBuffer)
        socket.getUnderlyingSocket().getOutputStream().flush()
      }
    } catch {
      return errSSLClosedAbort
    }
    return noErr
  }
  
  override func close() {
    if (isClosed()) {
      ObjC.throwException(JavaIoIOException(nsString: "Already closed"))
    }
    
    SSLClose(sslContext!)
    
    underlyingSocket?.close()
  }
  
  override func getInputStream() -> JavaIoInputStream! {
    return inputStream
  }
  
  override func getOutputStream() -> JavaIoOutputStream! {
    return outputStream
  }
}

class iOSSSLSocketFactory : JavaxNetSslSSLSocketFactory{
  override func createSocket(with s: JavaNetSocket!, with host: String!, with port: jint, withBoolean autoClose: jboolean) -> JavaNetSocket! {
    return iOSSSLSocket(underlyingSocket: s, hostName: host)
  }
  
  override func createSocket() -> JavaNetSocket! {
    // Not implemented
    return nil
  }
  
  override func createSocket(with host: String!, with port: jint) -> JavaNetSocket! {
    // Not implemented
    return nil
  }
  
  override func getSupportedCipherSuites() -> IOSObjectArray! {
    // Not implemented
    return nil
  }
  
  override func getDefaultCipherSuites() -> IOSObjectArray! {
    // Not implemented
    return nil
  }
}
```
</p>
</details>
<p></p>

Now, in theory, I guess this should work. Unfortunately, it only works sporadically: The handshake sometimes succeeds, but more often, `SSLHandshake` returns an error:

> -50 - One or more parameters passed to a function were not valid.

Interestingly, the code seems more likely to succeed if I log from within the callback functions or use breakpoints to stop execution, so my guess is there are some race conditions that I didn't think of. Is there something in the j2objc code that could cause these issues? Is there anything with the handling of the underlying Socket in the callbacks that I've overseen?

For testing the code, I use this simple test:
<details><summary>SSLSocketTest.java</summary>
<p>

```
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.InetSocketAddress;
import java.net.Socket;

import javax.net.ssl.SSLSocket;
import javax.net.ssl.SSLSocketFactory;

public class SSLSocketTest {
    private SSLSocketFactory mFactory;
    private String mHost;
    private int mPort;

    // For example: www.google.com, 443
    public SSLSocketTest(SSLSocketFactory factory, String host, int port){
        mFactory = factory;
        mHost = host;
        mPort = port;
    }

    public void run(){
        try {
            Socket socket = new Socket();
            socket.connect(new InetSocketAddress(mHost, mPort));
            SSLSocket sslSocket = (SSLSocket) mFactory.createSocket(socket, mHost, mPort, true);
            sslSocket.startHandshake();

            String request = "GET / HTTP/1.1\nHost:" + mHost + "\n\n";
            sslSocket.getOutputStream().write(request.getBytes());
            sslSocket.getOutputStream().flush();

            StringBuilder response = new StringBuilder();
            BufferedReader br = new BufferedReader(new InputStreamReader(sslSocket.getInputStream()));
            response.append(br.readLine());
            br.close();
        } catch (Exception e) {
            // Removed custom logging
        }
    }
}

```
</p>
</details>
<p></p>

For completeness:

<details><summary>ObjC.h</summary>
<p>

```
#import <Foundation/Foundation.h>

@interface ObjC : NSObject

+ (BOOL)catchException:(void(^)())tryBlock error:(__autoreleasing NSError **)error;
+ (void)throwException:(id)exception;

@end

```
</p>
</details>
<p></p>

<details><summary>ObjC.m</summary>
<p>

```
#import "ObjC.h"

@implementation ObjC

+ (BOOL)catchException:(void(^)())tryBlock error:(__autoreleasing NSError **)error {
  @try {
    tryBlock();
    return YES;
  }
  @catch (NSException *exception) {
    *error = [[NSError alloc] initWithDomain:exception.name code:0 userInfo:exception.userInfo];
    return NO;
  }
}

+ (void)throwException:(id)exception{
  @throw exception;
}

@end
```
</p>
</details>
<p></p>

Also, I'm wondering if an SSLSocketFactory like this could be added to j2objc? It seems like most of Java's SSLSocket functionality can be implemented with the Secure Transport API. Obviously that leaves things like certificate parsing to iOS (without possibility for any Java-side interventions), but I would think that's fine for most applications.

**Comment by tomball** [comment]
Right, j2objc's [IosSecurityProvider](https://github.com/google/j2objc/blob/master/jre_emul/Classes/com/google/j2objc/security/IosSecurityProvider.java) is missing a secure socket security provider or the cipher and certificate management providers it needs. The goal has always been to implement these providers using the iOS Security Framework and the Secure Transport API, since doing otherwise makes certificate management harder and risks opening security holes.

The reason this task is so hard is because the security and network APIs are so huge, which apps normally only need a narrow slice of, but each app seems to need something different. So each team hits this big support hole, then writes a little native code to address their needs and moves on. But maybe the solution is for each app to contribute their slice and the support gradually improves (that's how the HTTPS support was added).

So we'd love it if your team could [contribute](https://developers.google.com/j2objc/guides/how-to-contribute) a SSLSocketFactory implementation that uses the Secure Transport API, based on your WrappedSSLSocket class. It wouldn't need to be full-featured, just enough to meet the needs of your app. We can take Android's SSLSocketFactory unit tests and comment-out the ones that test unimplemented functionality. Then when someone complains about missing functionality, they can enhance the existing factory.

**Comment by FD-** [comment]
Alright, got it to work finally! The Secure Transport API really could deserve a better documentation! 

I'm willing to contribute this into your project, but my code is in Swift and I'm afraid I don't have the time to refresh my Objective C (any memory management). Could I just append the working Swift code here? Would you be willing to port that into a form that suits the project?

**Comment by tomball** [comment]
OK, but first please add a license comment to the top of each file (we use the Apache 2.0 license ([example](https://github.com/google/j2objc/blob/master/examples/Hello/HelloTest.m)), but pick the one you're most comfortable with), along with however Swift defines an @author tag. That way, it's obvious to anyone that you wrote the code and are willing to freely share it.

**Comment by FD-** [comment]
I've now created a repo and added all relevant code: https://github.com/FD-/SSLSocketFactory-j2objc. Please let me know if there's anything missing that you need for integrating it into j2objc!

**Comment by tomball** [comment]
Antonio, here is the project we discussed that implements SSLSocket using the iOS Secure Transport API. Thanks for the help, @FD-!

**Comment by antonio-cortes-perez** [comment]
perfect! I'll put it in my to-do list.

**Comment by FD-** [comment]
I  have found an issue with my implementation. In some rare cases, a read times out. I've investigated the issue and traced it down to what seems to be an issue with j2objc's Socket implementation: In some rare cases, the read on the underlying Socket never returns. It seems like at every occurrence of the issue, the SecureTransport API asked for 5 bytes from the underlying Socket. Is an issue in j2objc's Socket implementation possible?

**Comment by FD-** [comment]
From what I can tell by now, it seems like j2objc's Socket implementation doesn't properly return -1 when reaching EOF while reading. It might also be that the last buffer does not get passed to the application. I've yet to find a way to debug j2objc's SocketInputStream. I'll try to build a bug demo without the SSL layer.

**Comment by FD-** [comment]
Actually, the problem seems to be linked to chunked encoding over SSL. I haven't been able to reproduce the issue without the SSL layer. A URL that reproduces the issue is: https://jigsaw.w3.org/HTTP/ChunkedScript.

**Comment by FD-** [comment]
Perfect, thanks so much!

**Comment by FD-** [comment]
I finally managed to fix the issue. The problem mainly was that the SecureTransport API seems to request data from the underlying socket even if it still has data in its application data buffer. The remaining data from the buffer is only passed to the application after the callback reported that no more data from the underlying socket is currently available. This is problematic in our case here, as the underlying socket is blocking. We might end up having to wait until the read timeout from the underlying socket has expired, just to get the data that has been in the SSL buffer already. Once I had figured that out (Apple's documentation doesn't loose a word on it), it was pretty easy to fix the issue: the code now just checks if any data is available from the underlying socket, to make sure the read never blocks. I have updated my repo.

**Comment by tomball** [comment]
That's great! I'm glad you have a fix for your app, and appreciate you updating your repo.

**Comment by antonio-cortes-perez** [comment]
Hi @FD- , thanks for sharing your Swift implementation! 
Quick note, in iOSSSLOutputStream.flush, I think the synchronize should use self.sslSocket instead of only self.

**Comment by FD-** [comment]
Good catch! Fixed now in my repo.

**Comment by FD-** [comment]
Thanks for bringing the code into proper Objective C form! I read through your commits and noticed you removed the availability check in the read callback. Are you sure this still works reliably?

**Comment by antonio-cortes-perez** [comment]
Besides the unit tests, I used the code snippet below to guide the implementation. First, I made sure that it was a functional java program (javac/java). Then, my goal was to make the transpiled code work.

With the availability check, the program was hanging because SslReadCallback was not returning errSSLClosedGraceful. It looks like at the end of the stream, the availability check was always returning errSSLWouldBlock.

I just ran a script that calls 100 times the code below and it did not fail. I also used the URL that you mentioned above (https://jigsaw.w3.org/HTTP/ChunkedScript) and it seems to be working as expected.


  public static void main(String[] args) throws Exception {
    String host = "www.google.com";
    int port = 443;
    SSLSocketFactory factory = (SSLSocketFactory) SSLSocketFactory.getDefault();
    Socket socket = factory.createSocket(host, port);
    
    Writer out = new OutputStreamWriter(socket.getOutputStream());
    out.write("GET / HTTP/1.0\r\nHost: " + socket.getInetAddress().getHostName() + "\r\n\r\n");
    out.flush();

    BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream()));
    String line;
    String page = "";
    while ((line = in.readLine()) != null) {
      page += line;
    }
    assert page.matches(".*200 OK.*");
    System.out.println("OK");

    out.close();
    in.close();
    socket.close();
  }

**Comment by FD-** [comment]
I finally managed to build j2objc from source and unfortunately, I'm seeing problems with the new SSLSocketFactory. 
More specifically, I'm seeing timeouts (`java.net.SocketTimeoutException: Read timed out` and `java.net.SocketTimeoutException: timeout`) with connections that work with my SSLSocketFactory, but don't with the new implementation in j2objc. I think these are the same issues I worked around by adding my availability checks.

**Comment by FD-** [comment]
I agree that with the availability check, we don't get the correct closing behaviour. So, it seems like either implementation will timeout, just at different points. I'm going to revert to my implementation for now, as it seems to work with the rest of my project. I'm afraid I don't have the time for additional debugging atm.

**Comment by FD-** [comment]
Also, it seems like we need @autoreleasepools in the SSL callbacks. Otherwise, I'm seeing memory usage growing significantly (RAM usage grows faster than the file that is being downloaded).

**Comment by antonio-cortes-perez** [comment]
Thanks for the feedback! Do you have a code snippet that replicates the timeout? I would like to investigate it.

**Comment by FD-** [comment]
I've tried to isolate the issue, but wasn't really successful. So here is the most minimal setup that reproduces the bug:

- Use the OkHttp version from [here](https://github.com/FD-/OkHttp-j2objc). (This only contains minor modifications, so it shouldn't be the cause of the bug by itself)
- Request the content of `https://jigsaw.w3.org/HTTP/ChunkedScript`

This will fail with a timeout for ~50% of tries. 

Please note your tests with the chunked url likely were ineffective because the content is only served as chunks to HTTP 1.1 clients, while your test identified as HTTP 1.0.

**Comment by tomball** [comment]
[We set HTTP 1.1 here](https://github.com/google/j2objc/blob/master/jre_emul/Classes/com/google/j2objc/net/IosHttpURLConnection.java#L611), isn't that sufficient?

**Comment by FD-** [comment]
I'm a bit confused now: the code you linked seems to be from j2objc's HttpUrlConnection implementation, while I was referring to j2objc's SslSocket implementation.

**Comment by FD-** [comment]
I'm thinking it must be related to buffer sizes whether the issue shows or doesn't.

**Comment by tomball** [comment]
I assumed HttpURLConnection because SSLSocket doesn't have the notion of an
HTTP client. Maybe the MockWebServer we're using for tests needs to be
updated to report a 1.1 client, but I'm definitely not an expert.

**Comment by FD-** [comment]
Well, HTTP clients actually are just talking to an HTTP server via a socket. So, what we're doing here is just doing that manually, so we can see what's going through the socket.

What I meant was just that 

> out.write("GET / HTTP/1.0\r\nHost: " + socket.getInetAddress().getHostName() + "\r\n\r\n");

in the snippet above defines the client as HTTP/1.0. You just have to replace HTTP/1.0 with HTTP/1.1 and you're talking HTTP/1.1 :)

**Comment by FD-** [comment]
The strange thing is that:

 - On the one hand, it seems like OkHttp is the culprit: The same resource doesn't produce any timeout when requested 'manually' via sending the request string through the socket.
 - On the other hand, it seems like OkHttp should be working, as the same OkHttp source compiled for the desktop JRE loads the same resource just fine

**Comment by FD-** [comment]
Actually, I was just able to reproduce the problem without involvement of OkHttp, though with a url that isn't publicly accessible. 

My current theory: 

In some cases, the SSL layer doesn't forward the application data but accumulates it in its buffer to forward once the buffer is full. It seems like the data is also forwarded when the SSL layer gets notified that the server closed the connection. That's why your tests didn't reproduce the issue.

However, issues occur when the server doesn't close the connection once it sent all data. In that case, the application sometimes doesn't get the last (incomplete) buffer of application data. This is particularly problematic for http connections with chunked encoding, as it seems in that case, it's uncommon for the server to close the connection. Instead, the last buffer the server sends contains a chunk header of size 0, which tells the client that transfer finished and the connection can be closed.

Now that I think about it, this is actually the same conclusion I found when I investigated the issue a few week ago.

**Comment by FD-** [comment]
A possible solution would be to make the socket non-blocking temporarily. However, we have to be careful to still allow multi-layer ssl (ssl over ssl over ... over tcp).

**Comment by FD-** [comment]
I think I actually found a better solution: Instead of performing the availability check in the SSLReadCallback, we can use SecureTransport's SSLGetBufferedReadSize function to check whether there's data in the SSL buffer and make sure we deplete the buffer before the SSL layer issues a new read to the underlying socket.

Basically, I just modified the read method of the implementation:

```
- (jint)readWithByteArray:(IOSByteArray *)b
                  withInt:(jint)off
                  withInt:(jint)len {
  size_t processed = 0;
  OSStatus status;

  [_socket startHandshake];
  
  size_t available = 0;
  status = SSLGetBufferedReadSize(_socket->_sslContext, &available) != errSecSuccess;
  checkStatus(status);

  if (available != 0) {
    len = available;
  }
  
  @synchronized (_socket) {
    do {
      status = SSLRead(_socket->_sslContext, b->buffer_ + off, len, &processed);
    } while (status == errSSLWouldBlock && processed == 0);
  }

  if (status == errSSLClosedGraceful || status == errSSLClosedAbort) {
    processed = -1;
  } else {
    checkStatus(status);
  }
  if (_socket->_sslException) {
    @throw _socket->_sslException;
  }
  return (jint)processed;
}
```
Can probably implemented more efficiently.

Initial tests show that this shows expected behavior in all cases. Will further evaluate!

**Comment by antonio-cortes-perez** [comment]
Thanks for thorough investigation!

I was looking at the function st_SocketReadFunc in
https://www.videolan.org/developers/vlc/modules/misc/securetransport.c

And I noticed that when they read 0 bytes, they interpret that as EOF (errSSLClosedGraceful). I was wondering if this could also be a possible solution; right now, we are returning errSSLWouldBlock in this scenario.

**Comment by antonio-cortes-perez** [comment]
Also thanks for your suggestion about adding autoreleasepool. I just added them to the Ssl*Callback.

**Comment by FD-** [comment]
> Thanks for thorough investigation!
> 
> I was looking at the function st_SocketReadFunc in
> https://www.videolan.org/developers/vlc/modules/misc/securetransport.c
> 
> And I noticed that when they read 0 bytes, they interpret that as EOF (errSSLClosedGraceful). I was wondering if this could also be a possible solution; right now, we are returning errSSLWouldBlock in this scenario.

I just skimmed through the vlc code and AFAICT, this wouldn't work for us: VLC uses the readv call, which is clearly specified to return 0 in the case of EOF. In our case, EOF is reported as -1, and 0 can in theory be a valid result when no data is available atm, but might follow. As we allow any kind of underlying socket, I wouldn't recommend interpreting the result in a way that isn't immediately derived from the specification. It might work with a plain socket as the underlying socket, but might fail with other (custom) socket implementations (layers).

Actually, I like the solution using SSLGetBufferedReadSize, as described above, and so far, it seems to work. Also, I feel like this is how the SecureTransport API is meant to be used with blocking underlying channel.
Thus, I'd suggest to try it on your tests and see how it performs.

**Comment by antonio-cortes-perez** [comment]
Thanks for your insights! Your last two recommendations made it to the latest release.
Fixed in version 2.3.

**Comment by FD-** [comment]
Awesome!
